biz.aQute.bndlib/src/aQute/bnd/osgi/Analyzer.java

```diff
@@ -3708,4 +3708,63 @@ public boolean assignable(Clazz annoServiceClazz, Clazz inferredServiceClazz) {
 		return false;
 	}
 
+	/**
+	 * Guess the module name of the jar from
+	 * <p>
+	 * a) the module-info.class or,
+	 * <p>
+	 * b) the {@link Constants#AUTOMATIC_MODULE_NAME} manifest header or,
+	 * <p>
+	 * c) the {@link Constants#BUNDLE_SYMBOLICNAME} manifest header.
+	 *
+	 * @return null when the jar has no module-info.class, manifest, when the
+	 *         manifest has no {@link Constants#AUTOMATIC_MODULE_NAME} header
+	 *         and no {@link Constants#BUNDLE_SYMBOLICNAME} header, or when
+	 *         either header value is not a valid bsn according to {@link #BSN}.
+	 * @throws Exception when the jar is closed or when the manifest could not
+	 *             be retrieved.
+	 */
+	public String getModuleName(Jar jar) throws Exception {
```

Comment:
I guess I would ask why the method does not just return the module name for the jar being analyzed? That is, why to you need to pass a Jar object? Create an Analyzer with a Jar object (or setJar after construction) and then the getModuleName method takes no args. It just uses the set jar as the receiver.

Reply:
because I need to inspect every jar in the classpath to associate it's module name with a package it contains

Reply:
Hmm. This code parses the module-info.class file each time and then discards it. I suggested putting these methods in Analyzer since I had mistakenly assumed that the analyzer would be making the bundle for which you wanted this information and therefore the analyzer would already have the Clazz object (findClass). 

Given this information, we may then needs to have Jar do this which custom code which directly uses the ClassFIle class on module-info.class. It could then dig this information out when the Jar object is  built like we do for the manifest. WDYT? I could take a stab at that if you want.

Reply:
Sounds good!